design/baremetal-operator/bmc-events.md

```diff
@@ -0,0 +1,151 @@
+<!--
+ This work is licensed under a Creative Commons Attribution 3.0
+ Unported License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# Event Subscription API
+
+Users of bare metal hardware may want to receive events from the
+baseboard management controller (BMC) in order to act on them in the
+event of a hardware fault, increase in temperature, removal of a
+device, etc.
+
+## Status
+
+provisional
+
+## Summary
+
+The Redfish standard includes the ability to subscribe to events, which
+will cause hardware events to be sent in a particular format to a target
+URI. This design document describes a Metal3 API for configuring a
+subscription. While Redfish is the primary target for this design, the
+Ironic API is vendor-neutral and seeks to provide a unified interface
+for configuring events.
+
+## Motivation
+
+Some environments run workloads that need to deal with potential
+faults or environmental changes quicker than they would get an alert
+through other channels. For example, some workloads may have a sidecar
+container that knows how to deal with an alert that a particular network
+interface went down, or that the CPU temperature reached a certain
+threshold.
+
+### Goals
+
+- Provide an API to manage subscriptions to events
+
+### Non-Goals
+
+- [Configurable events and thresholds](#configurable-events-and-thresholds)
+- Any kind of event polling
+- Software for processing the events, i.e. any webhook
+- BMC's beyond Redfish for now
+
+## Proposal
+
+### User Stories
+
+- I'd like to configure a my BMC to send events to a target URL.
+- I'd like to filter the types of events the BMC sends to my target URL.
+- I'd like to provide context to a particular event subscription.
+- I'd like to provide arbitrary headers.
+- I'd like the baremetal-operator to reconcile on the
+  BMCEventSubscription resource, and ensure it's state is accurate in
+  Ironic.
+
+## Design Details
+
+### Implementation Details
+
+```yaml
+apiVersion: metal3.io/v1alpha1
+kind: BMCEventSubscription
+metadata:
+  name: worker-1-events
+spec:
+   hostRef: ostest-worker-1
+   targetURI: https://events.apps.corp.example.com/webhook
+   filters:
```

Comment:
Are the filter values common across vendors? Are they part of the Redfish standard or defined by Ironic or just free-form?

Reply:
I was looking at the values reported by Redfish and the ones reported by the HW vendor, what the HW vendor reports is a subset of what Redfish would accept.

http://redfish.dmtf.org/schemas/v1/Event.json#/definitions/EventType

Reply:
Do you have a link to the schema? This is more relevant to the Ironic design than here, we're an abstraction that sits on top so this is designed around their original proposal.

cc: @iurygregory

Reply:
http://redfish.dmtf.org/schemas/DSP2046_2021.1.html#eventtype

Reply:
@jzding I've checked a Dell R640 with FW `4.40.10.00` using `RedfishVersion:1.9.0`, the information from `EventService` (using v1.5.0 doesn't provide the `RegistryPrefix`/`RegistryPrefixes` so it's not supported) http://paste.openstack.org/show/806812/

Reply:
filters won't be necessary, since HW vendors may not have the support for `RegistryPrefix/RegistryPrefixes` we are not adding the `EventTypes` to ironic API, by default we will be using `Alert` option till HW vendors have the support for it.

For more details see the discussion: https://meetings.opendev.org/irclogs/%23openstack-ironic/%23openstack-ironic.2021-06-25.log.html#t2021-06-25T13:41:57

Reply:
I removed all references to filters.